### PR TITLE
Ibis4Test: Assemble dependencies

### PR DESCRIPTION
**Pull request by mhdirkse**
I tested this by running a command-line Maven build, and by running the Larva tests.

This introduces an error in Eclipse, in the new file deployTest/pom.xml, see below. This error seems to come from the dependency plugin MOJO. But the MOJO does not produce this error when Maven runs from the command line. This subproject is intended for CI/CD, not for usage in Eclipse. Therefore I propose we ignore this error.

Artifact has not been packaged yet. When used on reactor artifact, copy should be executed after packaging: see MDEP-187. 
 (org.apache.maven.plugins:maven-dependency-plugin:3.1.2:copy-dependencies:default:process-sources) 
 org.apache.maven.plugin.MojoExecutionException: Artifact has not been packaged yet. When used on reactor artifact, copy should be executed after 
 packaging: see MDEP-187. at org.apache.maven.plugins.dependency.AbstractDependencyMojo.copyFile(AbstractDependencyMojo.java:181) at 
 org.apache.maven.plugins.dependency.fromDependencies.CopyDependenciesMojo.copyArtifact(CopyDependenciesMojo.java:249) at 
 org.apache.maven.plugins.dependency.fromDependencies.CopyDependenciesMojo.doExecute(CopyDependenciesMojo.java:124) at 
 org.apache.maven.plugins.dependency.AbstractDependencyMojo.execute(AbstractDependencyMojo.java:144) at 
 org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137) at 
 org.eclipse.m2e.core.internal.embedder.MavenImpl.execute(MavenImpl.java:332) at 
 org.eclipse.m2e.core.internal.embedder.MavenImpl.lambda$8(MavenImpl.java:1379) at 
 org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:179) at 
 org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:114) at 
 org.eclipse.m2e.core.internal.embedder.MavenImpl.execute(MavenImpl.java:1378) at 
 org.eclipse.m2e.core.project.configurator.MojoExecutionBuildParticipant.build(MojoExecutionBuildParticipant.java:54) at 
 com.ianbrandt.tools.m2e.mdp.core.MdpBuildParticipant.executeMojo(MdpBuildParticipant.java:133) at 
 com.ianbrandt.tools.m2e.mdp.core.MdpBuildParticipant.build(MdpBuildParticipant.java:67) at 
 org.eclipse.m2e.core.internal.builder.MavenBuilderImpl.build(MavenBuilderImpl.java:135) at 
 org.eclipse.m2e.core.internal.builder.MavenBuilder$1.method(MavenBuilder.java:169) at 
 org.eclipse.m2e.core.internal.builder.MavenBuilder$1.method(MavenBuilder.java:1) at 
 org.eclipse.m2e.core.internal.builder.MavenBuilder$BuildMethod.lambda$1(MavenBuilder.java:114) at 
 org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:179) at 
 org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:114) at 
 org.eclipse.m2e.core.internal.builder.MavenBuilder$BuildMethod.lambda$0(MavenBuilder.java:105) at 
 org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:179) at 
 org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:153) at 
 org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:101) at 
 org.eclipse.m2e.core.internal.builder.MavenBuilder$BuildMethod.execute(MavenBuilder.java:88) at 
 org.eclipse.m2e.core.internal.builder.MavenBuilder.build(MavenBuilder.java:197) at 
 org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:832) at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45) at 
 org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:220) at 
 org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:263) at 
 org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:316) at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45) at 
 org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:319) at 
 org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:371) at 
 org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:392) at 
 org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:154) at 
 org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:244) at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)

### CODE REVIEW

File: deployTest/pom.xml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+							<goal>copy-dependencies</goal>
+						</goals>
+						<configuration>
+							<excludeTransitive>true</excludeTransitive>
+						</configuration>
+					</execution>
+				</executions>
+			</plugin>
+		</plugins>
+	</build>
```
Comment (nielsm5) [inline deployTest/pom.xml:56]: Kan je niet beter pom.type = war gebruiken?

File: test/pom.xml [hunk 2]
```diff
@@ -14,7 +14,6 @@
 	<properties>
 		<m2eclipse.wtp.contextRoot>iaf-test</m2eclipse.wtp.contextRoot>
 		<failOnMissingWebXml>false</failOnMissingWebXml>
-		<databaseHost>localhost</databaseHost>
 	</properties>
```
Comment (nielsm5) [inline test/pom.xml:17]: Deze moet hier niet weg, dan kunnen we iaf-test niet meer los draaien

File: deployTest/pom.xml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			<artifactId>geronimo-jms_1.1_spec</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>commons-dbcp</groupId>
+			<artifactId>commons-dbcp</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>commons-pool</groupId>
+			<artifactId>commons-pool</artifactId>
+		</dependency>
```
Comment (nielsm5) [inline deployTest/pom.xml:36]: uit eindelijk gaat het puur hier om, de dependencies voor de betreffende databases. 
Dit zou ook in de iaf test module kunnen. Er zit niet heel veel overhead/winst in het lostrekken van de profielen naar een andere module.

**Review by nielsm5 — CHANGES_REQUESTED** [review]

File: deployTest/pom.xml (hunk 1 above)
Comment (mhdirkse) [reply deployTest/pom.xml:56]: Bedoel je in plaats van dat ik transitieve dependencies weglaat? Ik denk het niet, want ik wil met dit project juist niet alleen de .war ophalen. Het gaat ook om de .jar files die ik hier als dependencies opgeef. Maar ik wil niet alle dependencies van het FF! neerzetten, want die zitten al in de .war.

Comment (mhdirkse) [reply deployTest/pom.xml]: Ik snap het. Ik had het project deployTest zo gemaakt dat hij de .war en losse .jar files naast elkaar zet in een subdirectory van target. Maar ik begrijp nu dat we gewoon de .war van ibis4test groter moeten maken. Daar heb ik net een commit voor gedaan.

File: deployTest/pom.xml (hunk 3 above)
Comment (mhdirkse) [reply deployTest/pom.xml]: Opgevolgd.

File: test/pom.xml (hunk 2 above)
Comment (mhdirkse) [reply test/pom.xml]: Dan moet ik hem weghalen uit de iaf/pom.xml, anders is hij dubbel.

Comment (mhdirkse) [reply test/pom.xml]: Gedaan.

**Comment by mhdirkse** [comment]
Zoals het geworden is in deze pull request, kan je de testen alleen draaien als je toegang hebt tot onze private repository. Het is veel moeilijker geworden om mee te ontwikkelen als je niet bij WeAreFrank! werkt. Je moet dan zorgen voor je eigen Nexus waarin je zelf de externe dependencies zet. Dan pas kan je je Maven build fixen.

Ik zie er nu een voordeel van om de externe dependencies buiten de .war te hebben. Dan kan iemand buiten WeAreFrank! de dependencies verzamelen en toevoegen, zonder dat Maven toegang nodig heeft tot die artifacts.

Ik denk dat deze pull request pas klaar is als we hierover hebben nagedacht en als we CONTRIBUTING.md hebben aangepast. Daarom is hij nu draft. Welke kant gaan we op?

**Comment by mhdirkse** [comment]
Voor de goede orde: ik heb deze versie niet getest. Enkele commits terug heb ik wel met succes de Larva tests gedraaid in Eclipse.

**Comment by mhdirkse** [comment]
This does not work properly yet. When I run in Eclipse, intending not to use the "oracle" profile, then the Oracle libraries are still included in the .war. I checked this in eclipse-workspace/.metadata/.plugins/org.eclipse.wst.server.core/tmp0/wtpwebapps/.

**Comment by mhdirkse** [comment]
No, my previous comment is incorrect. I tested successfully that I can create a regular .war without database drivers. It was a hiccup of Eclipse that this failed before. I fixed it by doing a server clean in Eclipse.

File: core/pom.xml [hunk 4]
```diff
@@ -585,7 +585,6 @@
 		<dependency>
 			<groupId>commons-dbcp</groupId>
 			<artifactId>commons-dbcp</artifactId>
-			<version>1.4</version>
 			<scope>test</scope>
 		</dependency>
```
Comment (gvanbrakel) [inline core/pom.xml:589]: Dbcp is niet nodig als dependency, kan hier helemaal verwijderd worden.

File: pom.xml [hunk 5]
```diff
@@ -207,6 +207,12 @@
         <artifactId>commons-net</artifactId>
         <version>1.4.1</version>
       </dependency>
+      <dependency>
+        <groupId>commons-dbcp</groupId>
+        <artifactId>commons-dbcp</artifactId>
+        <version>1.4</version>
+        <scope>provided</scope>
+      </dependency>
```
Comment (gvanbrakel) [inline pom.xml:215]: dbcp is niet nodig als dependency. Kan hier verwijderd worden.

File: test/pom.xml [hunk 6]
```diff
@@ -169,7 +167,33 @@
...
+				<dependency>
+					<groupId>commons-dbcp</groupId>
+					<artifactId>commons-dbcp</artifactId>
+					<scope>compile</scope>
+				</dependency>
+				<dependency>
+					<groupId>commons-pool</groupId>
+					<artifactId>commons-pool</artifactId>
+					<scope>compile</scope>
+				</dependency>
```
Comment (gvanbrakel) [inline test/pom.xml:195]: Deze dependencies zijn hier niet nodig

File: test/pom.xml [hunk 7]
```diff
@@ -169,7 +167,33 @@
...
+					<groupId>com.oracle.database.jdbc</groupId>
+					<artifactId>ojdbc8</artifactId>
+					<version>12.2.0.1</version>
+					<scope>compile</scope>
+				</dependency>
+				<dependency>
+					<groupId>org.apache.geronimo.specs</groupId>
+					<artifactId>geronimo-jms_1.1_spec</artifactId>
+					<scope>compile</scope>
+				</dependency>
```
Comment (gvanbrakel) [inline test/pom.xml:185]: JMS dependency hoort niet in het Oracle profile

Comment (mhdirkse) [reply test/pom.xml:190]: Ik heb een nieuw profiel "tomcat" gemaakt.

File: core/pom.xml (hunk 4 above)
Comment (mhdirkse) [reply core/pom.xml]: Gedaan.

File: pom.xml (hunk 5 above)
Comment (mhdirkse) [reply pom.xml]: Gedaan.

File: test/pom.xml (hunk 6 above)
Comment (mhdirkse) [reply test/pom.xml]: Gedaan.

**Comment by mhdirkse** [comment]
GitHub says that there is a conflict that has to be resolved. I am going to take care of that. I also think that I did not update CONTRIBUTING.md. I am going to investigate that as well

File: test/src/main/webapp/META-INF/context.xml [hunk 8]
```diff
@@ -12,7 +12,7 @@
 		name="jdbc/ibis4test-oracle"
 		type="javax.sql.DataSource"
 		driverClassName="oracle.jdbc.driver.OracleDriver"
-		url="jdbc:oracle:thin:@localhost:1521:xe"
+		url="jdbc:oracle:thin:@${databaseHost}:1521:xe"
```
Comment (gvanbrakel) [inline test/src/main/webapp/META-INF/context.xml:15]: Als je databaseHost ergens een waarde geeft (dat zie ik nergens gebeuren, overigens), dan kan je daar ``host.docker.internal`` voor gebruiken.

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: test/src/main/webapp/META-INF/context.xml (hunk 8 above)
Comment (mhdirkse) [reply test/src/main/webapp/META-INF/context.xml:15]: In test/pom.xml staat:

    &lt;databaseHost&gt;localhost&lt;/databaseHost&gt;

Ik denk dat dit in een eerdere commit is gebeurd die al gemerged is. Deze regel moet denk ik zo blijven, omdat je anders niet meer met Eclipse het FF! kan starten. In dat geval moet er `localhost` in de database URL staan.

File: example/pom.xml [hunk 9]
```diff
@@ -44,7 +44,6 @@
     <dependency>
       <groupId>commons-dbcp</groupId>
       <artifactId>commons-dbcp</artifactId>
-      <version>1.4</version>
     </dependency>
```
[the remark this answers is not in the document]
Comment (mhdirkse) [reply example/pom.xml]: Ik herinner mij dat Niels gezegd heeft dat hij hier moet. Er was iets bijzonders met Ibis4Example waardoor deze dependency nodig was. Niels wilde die situatie veranderen, waarna de dependency inderdaad weg kan. Maar ik weet niet of het al zover is.

File: test/src/main/webapp/META-INF/context.xml (hunk 8 above)
Comment (mhdirkse) [reply test/src/main/webapp/META-INF/context.xml:15]: Ik heb Gerrit gisteren gesproken over alle feedback items. Ik heb gehoord dat we geen handmatige acties willen vragen zoals ik gedaan heb door een README.md te schrijven. Gerrit heeft gezegd dat Maven dit werk moet doen, maar Niels heeft ooit gezegd dat we geen Docker containers willen starten vanuit Maven. Niels en Gerrit, kunnen jullie beslissen hoe het moet worden?

**Comment by mhdirkse** [comment]
Sorry, ik lette niet goed op en heb onbedoeld een conversation resolved. Ik kan niet meer vinden welke. Ik ben het overzicht kwijt welke conversaties nog relevant zijn.

**Review by gvanbrakel — CHANGES_REQUESTED** [review]
Martijn, 
Deze Pull Request is alweer een tijdje blijven liggen, en een beetje ingehaald door het voortschrijdende inzicht. Om er mee verder te kunnen kan je het beste kijken naar het WebSphere project. 
Ik denk dat de pom.xml voor Tomcat grotendeels hetzelfde kan zijn als die voor WebSphere. De dependencies zijn hetzelfde, alleen zal Tomcat webapp gebruiken in plaats van ear.
De plugin configuratie zal ook voor het grootste gedeelte hetzelfde zijn.

File: .gitignore [hunk 10]
```diff
@@ -11,4 +11,6 @@ bin
 .factorypath
 
 **/META-INF/maven
-**/META-INF/MANIFEST.MF
\ No newline at end of file
+**/META-INF/MANIFEST.MF
+
+test/src/main/tools/appServers/tomcat/iaf-test.war
```
[the remark this answers is not in the document]
Comment (mhdirkse) [reply .gitignore]: Ik haal het weg.

File: core/pom.xml [hunk 11]
```diff
@@ -576,6 +576,12 @@
 			<version>4.12</version>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.hsqldb</groupId>
+			<artifactId>hsqldb</artifactId>
+			<version>2.3.0</version>
+			<scope>test</scope>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (mhdirkse) [reply core/pom.xml:579]: Ik ga hem weghalen.

**Comment by mhdirkse** [comment]
Inmiddels is er een andere pull request gemaakt om Docker images te maken met ibis4test. Ik sluit deze pull request.

**Pull request closed without merging**